src/ripple/app/tx/impl/Escrow.cpp

```diff
@@ -90,39 +90,42 @@ static inline bool after (NetClock::time_point now, std::uint32_t mark)
 }
 
 XRPAmount
-EscrowCreate::calculateMaxSpend(STTx const& tx)
+EscrowCreate::calculateMaxXRPSpend(STTx const& tx)
 {
-    return tx[sfAmount].xrp();
+    auto const& amount {tx[sfAmount]};
+    return amount.native() && !amount.negative() ?
+        amount.xrp() : beast::zero;
```

Comment:
Somewhat surprisingly, it turns out that the check is necessary and it is not okay to add an `assert` here.  The issue is that `Consequences` are returned from `preflight()` unconditionally.  So no matter what error code `preflight()` returns, and even if `preflight()` detects a bad `amount`, `makeTxConsequences()`will be called.  Therefore `makeTxConsequences()` must be prepared to handle bad values.

Reply:
It could be done, but I'm not convinced the effort and complexity would pay off.  Here's where `TxConsequences` are being collected now: https://github.com/ripple/rippled/blob/40c9999d18e66438e47ff7d76959b58d50ba2c3d/src/ripple/app/tx/impl/applySteps.cpp#L48-L73

So the `TxConsequences` are computed unconditionally.  At the moment we don't even know which is called first: `T::preflight` or the `TxConsequences` constructor.

To change this we'd change the three `invoke_preflight_helper` functions to call `preflight` first and, based on the return code, call one of two `TxConsequences` constructors.  That's certainly doable.

What I'm uncertain of is the benefit.  To keep `TxConsequences` low cost we'd like them to have half-way reasonable responses to all the questions they can answer.  The easiest way to get that answer is to consult the actual transaction in question.

Maybe I can turn the question around.  What would you expect a `TxConsequences` to look like when `preflight` returns a `tel` code?  Is that something that can be generalized?  Perhaps we should be returning `optional<TxConsequences>`?